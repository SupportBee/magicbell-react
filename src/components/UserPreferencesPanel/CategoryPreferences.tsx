import { useNotificationPreferences } from '@magicbell/react-headless';
import { capitalize, kebabCase } from 'lodash';
import React from 'react';
import ToggleInput from './ToggleInput';

interface Props {
  category: string;
}

export default function CategoryPreferences({ category }: Props) {
  const preferences = useNotificationPreferences();
  const categoryId = kebabCase(category);
  const categoryTitle = capitalize(categoryId.replace(/-/g, ' '));

  const updatePreferences = (data) => {
    preferences.save({ categories: { [category]: data } });
  };

  return (
    <>
      <div>{categoryTitle}</div>
      <div>
        <ToggleInput
          id={`${categoryId}-inapp`}
          value={preferences.categories[category].inApp}
          onClick={(value) => updatePreferences({ inApp: value })}
        />
      </div>
      <div>
        <ToggleInput
          id={`${categoryId}-email`}
          value={preferences.categories[category].email}
          onClick={(value) => updatePreferences({ email: value })}
        />
      </div>
      <div>
        <ToggleInput
          id={`${categoryId}-web-push`}
          value={preferences.categories[category].webPush}
          onClick={(value) => updatePreferences({ webPush: value })}
        />
      </div>
    </>
  );
}
