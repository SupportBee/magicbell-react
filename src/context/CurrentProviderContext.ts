import { createContext } from 'react';

export const NO_PROVIDER = 'MAGICBELL_NO_PROVIDER';

const CurrentProviderContext = createContext<String>(NO_PROVIDER);
export default CurrentProviderContext;
