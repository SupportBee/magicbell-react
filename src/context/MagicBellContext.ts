import { createContext, useContext } from 'react';

export interface IMagicBellContext {
  images?: Partial<{
    emptyInboxUrl: string;
  }>;
}

const MagicBellContext = createContext<IMagicBellContext>({} as IMagicBellContext);
export default MagicBellContext;

const useMagicBellContext = () => useContext(MagicBellContext);
export { useMagicBellContext };
