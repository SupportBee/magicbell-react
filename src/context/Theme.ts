export interface IElementTheme {
  backgroundColor: string;
  backgroundOpacity: number;
  borderRadius: string;
  fontFamily: string;
  fontSize: string;
  textAlign: 'left' | 'right' | 'center' | 'justify' | 'initial' | 'inherit';
  textColor: string;
  textTransform: 'none' | 'capitalize' | 'uppercase' | 'lowercase' | 'initial' | 'inherit';
}

export interface IMagicBellTheme {
  icon: {
    borderColor: string;
    width: string;
  };
  unseenBadge: IElementTheme;
  header: IElementTheme;
  container: IElementTheme;
  footer: IElementTheme;
  notification: {
    default: IElementTheme;
    unread: IElementTheme;
    unseen: IElementTheme;
  };
}

const blue = '#3498F4';
const red = '#DF4759';
const black = '#3A424D';

// prettier-ignore
const fontFamily = '-apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen-Sans, Ubuntu, Cantarell, "Helvetica Neue", sans-serif';
const fontSize = '14px';
const textAlign = 'left';
const textTransform = 'none';

export const defaultTheme = {
  icon: {
    borderColor: blue,
    width: '24px',
  },
  header: {
    backgroundColor: blue,
    backgroundOpacity: 1,
    borderRadius: '8px',
    fontFamily,
    fontSize,
    textAlign,
    textColor: 'white',
    textTransform,
  },
  footer: {
    backgroundColor: blue,
    backgroundOpacity: 1,
    borderRadius: '8px',
    fontFamily,
    fontSize: fontSize,
    textAlign,
    textColor: 'white',
    textTransform,
  },
  unseenBadge: {
    backgroundColor: red,
    backgroundOpacity: 1,
    borderRadius: '2px',
    fontFamily,
    fontSize: '10px',
    textAlign: 'center',
    textColor: 'white',
    textTransform,
  },
  container: {
    backgroundColor: '#FFFFFF',
    backgroundOpacity: 1,
    borderRadius: '8px',
    fontFamily,
    fontSize: '14px',
    textAlign,
    textColor: black,
    textTransform,
  },
  notification: {
    default: {
      backgroundColor: blue,
      backgroundOpacity: 0.1,
      borderRadius: '8px',
      fontFamily,
      fontSize,
      textAlign,
      textColor: black,
      textTransform,
    },
    unread: {
      backgroundColor: blue,
      backgroundOpacity: 0,
      borderRadius: '8px',
      fontFamily,
      fontSize,
      textAlign,
      textColor: black,
      textTransform,
    },
    unseen: {
      backgroundColor: blue,
      backgroundOpacity: 0.05,
      borderRadius: '8px',
      fontFamily,
      fontSize,
      textAlign,
      textColor: black,
      textTransform,
    },
  },
} as IMagicBellTheme;
