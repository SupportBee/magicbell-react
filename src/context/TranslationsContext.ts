import { createContext, useContext } from 'react';

const TranslationsContext = createContext<any>({});
export default TranslationsContext;

const TranslationsProvider = TranslationsContext.Provider;
const useTranslations = () => useContext(TranslationsContext);
export { TranslationsProvider, useTranslations };
