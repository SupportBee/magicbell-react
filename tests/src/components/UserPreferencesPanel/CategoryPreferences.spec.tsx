import { useNotificationPreferences } from '@magicbell/react-headless';
import { fireEvent, render, RenderResult } from '@testing-library/react';
import { renderHook } from '@testing-library/react-hooks';
import React from 'react';
import CategoryPreferences from '../../../../src/components/UserPreferencesPanel/CategoryPreferences';
import { MagicBellThemeProvider } from '../../../../src/context/MagicBellThemeContext';
import { defaultTheme } from '../../../../src/context/Theme';

describe('components', () => {
  describe('CategoryPreferences', () => {
    let view: RenderResult;

    beforeEach(() => {
      useNotificationPreferences.setState({
        categories: {
          comments: { email: false, webPush: true, inApp: true },
          new_reply: { email: true, webPush: false, inApp: false },
        },
        lastFetchedAt: Date.now(),
      });

      view = render(
        <MagicBellThemeProvider value={defaultTheme}>
          <CategoryPreferences category="comments" />
        </MagicBellThemeProvider>,
      );
    });

    afterEach(() => {
      view.unmount();
    });

    describe('render', () => {
      it('renders a checkbox for each channel', () => {
        expect(view.container).toMatchSnapshot();
      });
    });

    describe('.updatePreferences', () => {
      it('updates the preferences for the inApp category', async () => {
        const { result } = renderHook(() => useNotificationPreferences());
        const spy = jest.spyOn(result.current, 'save');

        const inAppCheckbox = view.getAllByRole('checkbox')[0];
        fireEvent.click(inAppCheckbox);

        expect(spy).toHaveBeenCalledTimes(1);
        expect(result.current.categories).toMatchObject({ comments: { inApp: false } });
      });

      it('updates the preferences for the email category', async () => {
        const { result } = renderHook(() => useNotificationPreferences());
        const spy = jest.spyOn(result.current, 'save');

        const emailCheckbox = view.getAllByRole('checkbox')[1];
        fireEvent.click(emailCheckbox);

        expect(spy).toHaveBeenCalledTimes(1);
        expect(result.current.categories).toMatchObject({ comments: { email: true } });
      });

      it('updates the preferences for the webPush category', async () => {
        const { result } = renderHook(() => useNotificationPreferences());
        const spy = jest.spyOn(result.current, 'save');

        const webPushCheckbox = view.getAllByRole('checkbox')[2];
        fireEvent.click(webPushCheckbox);

        expect(spy).toHaveBeenCalledTimes(1);
        expect(result.current.categories).toMatchObject({ comments: { webPush: false } });
      });
    });
  });
});
