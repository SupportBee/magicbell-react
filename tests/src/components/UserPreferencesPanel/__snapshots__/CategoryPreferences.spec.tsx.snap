// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components CategoryPreferences render renders a checkbox for each channel 1`] = `
<div>
  <div>
    Comments
  </div>
  <div>
    <div>
      <input
        checked=""
        id="comments-inapp"
        type="checkbox"
      />
    </div>
  </div>
  <div>
    <div>
      <input
        id="comments-email"
        type="checkbox"
      />
    </div>
  </div>
  <div>
    <div>
      <input
        checked=""
        id="comments-web-push"
        type="checkbox"
      />
    </div>
  </div>
</div>
`;
